pkg/reconciler/topology/partitionset_reconcile.go

```diff
@@ -0,0 +1,278 @@
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"old partition could not get deleted",
+				)
+				return err
+			}
+		}
+	}
+	// Create partitions when no existing partition for the set has the same selector.
+	for key, matchLabels := range matchLabelsMap {
+		if _, ok := existingMatches[key]; !ok {
+			partition := generatePartition(partitionSet.Name, newMatchExpressions, matchLabels, partitionSet.Spec.Dimensions)
+			partition.OwnerReferences = []metav1.OwnerReference{
+				*metav1.NewControllerRef(partitionSet, topologyv1alpha1.SchemeGroupVersion.WithKind("PartitionSet")),
+			}
+			_, err = c.createPartition(ctx, logicalcluster.From(partitionSet).Path(), partition)
+			if err != nil && !apierrors.IsAlreadyExists(err) {
+				conditions.MarkFalse(
+					partitionSet,
+					topologyv1alpha1.PartitionsReady,
+					topologyv1alpha1.ErrorGeneratingPartitionsReason,
+					conditionsv1alpha1.ConditionSeverityError,
+					"partition could not get created",
+				)
+				return err
+			}
+		}
+	}
+	conditions.MarkTrue(partitionSet, topologyv1alpha1.PartitionsReady)
+	return nil
+}
+
+// partition populates shard label selectors according to dimensions.
+// It only keeps selectors that have at least one match.
+func partition(shards []*corev1alpha1.Shard, dimensions []string, shardSelectorLabels map[string]string) (matchLabelsMap map[string]map[string]string) {
+	matchLabelsMap = make(map[string]map[string]string)
+	labels := make([]string, len(dimensions), len(dimensions)+len(shardSelectorLabels))
+	copy(labels, dimensions)
+	for label := range shardSelectorLabels {
+		labels = append(labels, label)
+	}
+	// Sorting for consistent comparison.
+	sort.Strings(labels)
+	for _, shard := range shards {
+		key := ""
+		selector := make(map[string]string)
+		for _, label := range labels {
+			labelValue, ok := shard.Labels[label]
+			if !ok {
+				break
+			}
+			key = key + "+" + label + "=" + labelValue
+			selector[label] = labelValue
+		}
+		if len(key) > 0 {
+			matchLabelsMap[key] = selector
+		}
+	}
+	return matchLabelsMap
+}
+
+// isSelectorWithoutConflict checks for collisions between matchLabels inherited from shardSelector
+// and the ones generated from dimensions. MatchLabels are a map, which can only contain a single
+// occurrence of a key.
+func isSelectorWithoutConflict(selector *metav1.LabelSelector, dimensions []string) (bool, error) {
+	if selector == nil || selector.MatchLabels == nil {
+		return true, nil
+	}
+	for _, dimension := range dimensions {
+		if _, ok := selector.MatchLabels[dimension]; ok {
+			return false, fmt.Errorf("MatchLabel with key %s conflicts with a dimension, please use matchExpressions instead", dimension)
+		}
+	}
+	return true, nil
+}
+
+func (c *controller) deletePartitions(ctx context.Context, partitionSet *topologyv1alpha1.PartitionSet) error {
+	partitionSet.Status.Count = 0
```

Comment:
oh, it tracks the number of associated `partitions`.
should we then decrease this number by the number of successful calls to `deletePartition` ?

Reply:
`deletePartitions` (mind the `s` at the end) is only call in case of an issue with the selector (lines 55 and 83) the count should then be set back to 0. In case of regular deletion because of a change in shards `deletePartition`without `s` is called (lines 147 and 182). In case of no error with the selector the number of endpoints is already captured in the size of `matchLabelsMap`, no need for additional counting.
I will move `partitionSet.Status.Count = 0`one level up as it is not clear from the name of the function that it does that as well.